### PR TITLE
 Add unit test ensuring connections are created per-host 

### LINKED ISSUE #415: How are requests with differing `Host:` header field values mapped to outbound connections?
This is a follow-up to PR #397, in particular https://github.com/runconduit/conduit/pull/397#discussion_r169507498.

In the case where we send the request to the `SO_ORIGINAL_DST` destination, because we didn't discover the destination using the Destinations service, we're not making anything w.r.t. whether requests with different `Host` request header field values are sent on the same connection; the original service is making the choice whether or not to do that. We should, at least, document this in the code with a comment.

In the case where HTTP/1.x requests with differing host values are inbound to the proxy from an external source:
* What do we currently do?
* What *should* we do? In particular, should we segregate such requests by Host header field value, and have (at least) one connection to the proxied service per Host field value?

In the case where HTTP/1.x requests with different host values are outbound from the proxy to an external source:
* What do we currently do?
* What *should* we do? Again, should we segregate such requests by Host header field value, and have (at least) one connection to the external service per Host header field value?

Note in particular that in the outbound case we don't know whether the destination supports HTTP/1.1 or only HTTP/1.0; i.e. we don't know whether it might ignore the Host header. Also, even in the case where the destination claims to support HTTP/1.1, I'm not sure we can rely on this to mean that we can safely send it requests with differing Host header field values on the same connection. We should do what web browsers like Chrome do here.

All of the above is regarding plaintext (http://) connections. In the case of TLS-protected (https://) connections, I already know that at least empirical results from browsers require that the Host header field match the SNI (Server Name Indication) TLS extension. That is, HTTPS rules require us to avoid sending requests with different Host header field values on the same connection. See https://bugs.chromium.org/p/chromium/issues/detail?id=615413#c2.

Unless there's important reasons to do otherwise, I'd prefer to do the same for HTTP as is required for HTTPS; i.e. have distinct connections per Host request field value.

### PR DESCRIPTION
**Pull request by hawkw**
This PR adds a test that if multiple Hosts correspond to the same server, HTTP/1.1 requests with different Hosts will have separate connections. This already happens as a result of Hyper's connection pooling behaviour, but the unit test will ensure that it _continues_ to do so.

This requires @seanmonstar's  `proxy-te` branch, so we should merge this branch after we merge #457

### CODE REVIEW

File: proxy/tests/support/server.rs [hunk 1]
```diff
@@ -30,6 +30,13 @@ pub struct Server {
 pub struct Listening {
     pub addr: SocketAddr,
     pub(super) shutdown: Shutdown,
+    pub(super) conn_count: Arc<Mutex<usize>>,
```
Comment (seanmonstar) [inline proxy/tests/support/server.rs]: Prefer an `AtomicUsize` in almost any case where you want a `Mutex<usize>`.

Comment (hawkw) [reply proxy/tests/support/server.rs]: Done! 954d64193a6e1025d95a633cfe988b9551fef503

**Comment by hawkw** [comment]
@briansmith and @seanmonstar, I'd love your reviews on this PR when you have a chance.

**Comment by briansmith** [comment]
I'm not sure if more work is going to be done here. Based on https://github.com/runconduit/conduit/issues/415#issuecomment-369434644 I would expect at least one more test where there is no Host header at all, and some code changes to deal with that case, if this is really supposed to close #415.

**Comment by hawkw** [comment]
@briansmith I can add the additional test as well. When @seanmonstar mentioned the need to handle the case where there is no `Host:` header and no authority, I thought I'd open a PR for the branch I wrote earlier to change the behavior as well, since it will ensure we handle that case correctly.

**Comment by briansmith** [comment]
> @briansmith I can add the additional test as well. When @seanmonstar mentioned the need to handle the case where there is no Host: header and no authority, I thought I'd open a PR for the branch I wrote earlier to change the behavior as well, since it will ensure we handle that case correctly.

OK, then let's remove "Closes #415" from this PR. I'll review it today.

**Comment by hawkw** [comment]
> OK, then let's remove "Closes #415" from this PR. I'll review it today.

Sounds good, I'll let you know once I've added the additional unit test & you can review then.

**Comment by hawkw** [comment]
Okay, 6ee8042f07bbf89d61a2b58e76142bf1e7cc8d67 adds a test for the handling of requests with no Host header. However, it fails against the current proxy (as expected), so the test will fail on this branch.

Therefore, I'm going to open a new PR for my changes that add a `Host` key to `Recognize`, and merge these tests into that branch, so that we can merge the tests alongside the changes that should make them pass.

**Comment by briansmith** [comment]
Closing this on the understanding taht #492 supercedes this, since that's what the description in #492 says.

**Pull request closed without merging**